File: Codigo/back-end/src/main/java/br/pucminas/graphtest/adapters/inbound/dto/decisiontable/DecisionTableDTO.java
package br.pucminas.graphtest.adapters.inbound.dto.decisiontable;

import br.pucminas.graphtest.application.domain.decisiontable.enums.DecisionTableCellValueEnum;
import br.pucminas.graphtest.application.domain.decisiontable.enums.DecisionTableSyncStatusEnum;
import lombok.Builder;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Builder
public record DecisionTableDTO(
        UUID id,
        UUID gceId,
        UUID projectId,
        String name,
        String description,
        String sourceFingerprint,
        DecisionTableSyncStatusEnum syncStatus,
        LocalDateTime sourceGceUpdatedAt,
        LocalDateTime createdAt,
        LocalDateTime updatedAt,
        List<DecisionTableConditionDTO> conditions,
        List<DecisionTableActionDTO> actions,
        List<DecisionTableRuleDTO> rules,
        List<DecisionTableConditionCellDTO> conditionCells,
        List<DecisionTableActionCellDTO> actionCells
) {

    public record DecisionTableConditionDTO(
            UUID id,
            UUID decisionTableId,
            String code,
            String label,
            int orderIndex,
            LocalDateTime createdAt,
            LocalDateTime updatedAt
    ) {
    }

    public record DecisionTableActionDTO(
            UUID id,
            UUID decisionTableId,
            String code,
            String label,
            int orderIndex,
            LocalDateTime createdAt,
            LocalDateTime updatedAt
    ) {
    }

    public record DecisionTableRuleDTO(
            UUID id,
            UUID decisionTableId,
            String code,
            String description,
            int orderIndex,
            LocalDateTime createdAt,
            LocalDateTime updatedAt
    ) {
    }

    public record DecisionTableConditionCellDTO(
            UUID id,
            UUID ruleId,
            UUID conditionId,
            DecisionTableCellValueEnum value,
            LocalDateTime createdAt,
            LocalDateTime updatedAt
    ) {
    }

    public record DecisionTableActionCellDTO(
            UUID id,
            UUID ruleId,
            UUID actionId,
            DecisionTableCellValueEnum value,
            LocalDateTime createdAt,
            LocalDateTime updatedAt
    ) {
    }
}

File: Codigo/front-end/src/features/decision-table/types/decisionTableDTO.ts
export type DecisionTableSyncStatus = 'UP_TO_DATE' | 'STALE';
export type BackendConditionValue = 'YES' | 'NO' | 'IRRELEVANT';
export type BackendActionValue = 'YES' | 'NO';

export interface DecisionTableConditionDTO {
  id: string;
  decisionTableId: string;
  code: string;
  label: string;
  orderIndex: number;
  createdAt: string;
  updatedAt: string;
}

export interface DecisionTableActionDTO {
  id: string;
  decisionTableId: string;
  code: string;
  label: string;
  orderIndex: number;
  createdAt: string;
  updatedAt: string;
}

export interface DecisionTableRuleDTO {
  id: string;
  decisionTableId: string;
  code: string;
  description: string;
  orderIndex: number;
  createdAt: string;
  updatedAt: string;
}

export interface DecisionTableConditionCellDTO {
  id: string;
  ruleId: string;
  conditionId: string;
  value: BackendConditionValue;
  createdAt: string;
  updatedAt: string;
}

export interface DecisionTableActionCellDTO {
  id: string;
  ruleId: string;
  actionId: string;
  value: BackendActionValue;
  createdAt: string;
  updatedAt: string;
}

export interface DecisionTableDTO {
  id: string;
  gceId: string;
  projectId: string;
  name: string;
  description: string;
  sourceFingerprint: string;
  syncStatus: DecisionTableSyncStatus;
  sourceGceUpdatedAt: string;
  createdAt: string;
  updatedAt: string;
  conditions: DecisionTableConditionDTO[];
  actions: DecisionTableActionDTO[];
  rules: DecisionTableRuleDTO[];
  conditionCells: DecisionTableConditionCellDTO[];
  actionCells: DecisionTableActionCellDTO[];
}

export interface UpdateDecisionTableDetailsDTO {
  name: string;
  description?: string;
}

// Assinatura de teste funcional (gerada a partir da tabela de decisão).
export interface FunctionalTestConditionDTO {
  conditionId: string;
  code: string;
  label: string;
  value: BackendConditionValue;
}

export interface FunctionalTestActionDTO {
  actionId: string;
  code: string;
  label: string;
  value: BackendActionValue;
}

export interface FunctionalTestMethodSignatureDTO {
  ruleId: string;
  ruleCode: string;
  methodName: string;
  conditions: FunctionalTestConditionDTO[];
  actions: FunctionalTestActionDTO[];
  generatedCode: string;
}

export interface GenerateFunctionalTestSignatureResponseDTO {
  decisionTableId: string;
  gceId: string;
  projectId: string;
  decisionTableName: string;
  rulesCount: number;
  testMethods: FunctionalTestMethodSignatureDTO[];
  generatedCode: string;
  warnings: string[];
}
